endpoints/openrtb2/amp_auction.go

```diff
@@ -276,6 +278,94 @@ func (deps *endpointDeps) loadRequestJSONForAmp(httpRequest *http.Request) (req
 		*req.Imp[0].Secure = 1
 	}
 
+	// Override query params
+	if width, err := strconv.ParseUint(httpRequest.FormValue("w"), 10, 64); err == nil {
+		if req.Imp[0].Banner != nil {
+
+			*req.Imp[0].Banner.W = width
+		}
+	}
+
+	if height, err := strconv.ParseUint(httpRequest.FormValue("h"), 10, 64); err == nil {
+		if req.Imp[0].Banner != nil {
+			*req.Imp[0].Banner.H = height
+		}
+	}
+
+	if overrideWidth, err := strconv.ParseUint(httpRequest.FormValue("ow"), 10, 64); err == nil {
+		// override w
+		if req.Imp[0].Banner != nil {
+			*req.Imp[0].Banner.W = overrideWidth
+		}
+	}
+
+	if overrideHeight, err := strconv.ParseUint(httpRequest.FormValue("oh"), 10, 64); err == nil {
+		// override h
+		if req.Imp[0].Banner != nil {
+			*req.Imp[0].Banner.H = overrideHeight
+		}
+	}
+
+	multiSize := httpRequest.FormValue("ms")
+	if multiSize != "" {
+		// override w and h with format
+		req.Imp[0].Banner.H = nil
+		req.Imp[0].Banner.W = nil
+		sizes := strings.Split(multiSize, ",")
+
+		format := make([]openrtb.Format, len(sizes))
+		for _, size := range sizes {
+			wh := strings.Split(size, "x")
+			if len(wh) == 2 {
+				f := openrtb.Format{}
+				if width, err := strconv.ParseUint(wh[0], 10, 64); err == nil {
+					f.W = width
+				} else {
+					continue
+				}
+				if height, err := strconv.ParseUint(wh[1], 10, 64); err == nil {
+					f.H = height
+				} else {
+					continue
+				}
+
+				format = append(format, f)
+			}
+		}
+		req.Imp[0].Banner.Format = format
+	}
+
+	slot := httpRequest.FormValue("slot")
+	if slot != "" {
+		// ad server ad unit ID?
+	}
+
+	targeting := httpRequest.FormValue("targeting")
+	if targeting != "" {
+		// TODO: should parse to a JSON object with KV targeting pairs
+		// Should this go in site.keywords or user.keywords?
+	}
+
+	pageURL := httpRequest.FormValue("purl")
+	if pageURL != "" {
+		req.Site.Page = pageURL
+	}
+
+	canonicalURL := httpRequest.FormValue("curl")
+	if canonicalURL != "" {
+		// override purl
+		req.Site.Page = canonicalURL
+	}
+
+	if timeout, err := strconv.ParseInt(httpRequest.FormValue("timeout"), 10, 64); err == nil {
+		req.TMax = timeout
```

Comment:
Do we want to use _exactly_ the AMP timeout here, or should we subtract off some time for the estimated latency?

I know we can't really estimate the latency in any quantitative way... but it seems like pubs would see much better performance overall if we did _something_ here. If our auctions time out at the same time the AMP call does, we'll probably miss a lot of bids.

Maybe worth knocking off 10-20ms here? Or make it configurable so that PBS hosts can decide based on their normal user geos & datacenter spread.